pkg/fleetautoscalers/controller.go

```diff
@@ -0,0 +1,354 @@
+	if !cache.WaitForCacheSync(stop, c.fleetAutoScalerSynced) {
+		return errors.New("failed to wait for caches to sync")
+	}
+
+	c.workerqueue.Run(workers, stop)
+	return nil
+}
+
+// creationMutationHandler will intercept when a FleetAutoScaler is created, and attach it to the requested Fleet
+// assuming that it exists. If not, it will reject the AdmissionReview.
+func (c *Controller) creationMutationHandler(review admv1beta1.AdmissionReview) (admv1beta1.AdmissionReview, error) {
+	c.logger.WithField("review", review).Info("creationMutationHandler")
+	obj := review.Request.Object
+	fas := &stablev1alpha1.FleetAutoScaler{}
+	err := json.Unmarshal(obj.Raw, fas)
+	if err != nil {
+		return review, errors.Wrapf(err, "error unmarshalling original FleetAutoScaler json: %s", obj.Raw)
+	}
+
+	var causes []metav1.StatusCause
+	causes = fas.ValidateAutoScalingSettings(causes)
+	if len(causes) != 0 {
+		review.Response.Allowed = false
+		details := metav1.StatusDetails{
+			Name:   review.Request.Name,
+			Group:  review.Request.Kind.Group,
+			Kind:   review.Request.Kind.Kind,
+			Causes: causes,
+		}
+		review.Response.Result = &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Message: "FleetAutoScaler creation is invalid",
+			Reason:  metav1.StatusReasonInvalid,
+			Details: &details,
+		}
+		return review, nil
+	}
+
+	// When being called from the API the fas.ObjectMeta.Namespace isn't populated
+	// (whereas it is from kubectl). So make sure to pull the namespace from the review
+	fleet, err := c.fleetLister.Fleets(review.Request.Namespace).Get(fas.Spec.FleetName)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			review.Response.Allowed = false
+			review.Response.Result = &metav1.Status{
+				Status:  metav1.StatusFailure,
+				Reason:  metav1.StatusReasonInvalid,
+				Message: "Invalid FleetAutoScaler",
+				Details: &metav1.StatusDetails{
+					Name:  review.Request.Name,
+					Group: review.Request.Kind.Group,
+					Kind:  review.Request.Kind.Kind,
+					Causes: []metav1.StatusCause{
+						{Type: metav1.CauseTypeFieldValueNotFound,
+							Message: fmt.Sprintf("Could not find fleet %s in namespace %s", fas.Spec.FleetName, review.Request.Namespace),
+							Field:   "fleetName"}},
+				},
+			}
+			return review, nil
+		}
+		return review, errors.Wrapf(err, "error retrieving fleet %s", fas.Name)
+	}
+
+	// Attach to the fleet
+	// When a Fleet is deleted, the FleetAutoScaler should go with it
+	ref := metav1.NewControllerRef(fleet, stablev1alpha1.SchemeGroupVersion.WithKind("Fleet"))
+	fas.ObjectMeta.OwnerReferences = append(fas.ObjectMeta.OwnerReferences, *ref)
+
+	// Q: can we do this nicer?
+	c.recorder.Eventf(fleet, corev1.EventTypeNormal, string("CreatingFleetAutoScaler"), "Attached FleetAutoScaler %s to fleet %s", fas.ObjectMeta.Name, fas.Spec.FleetName)
+	c.recorder.Eventf(fas, corev1.EventTypeNormal, string("CreatingFleetAutoScaler"), "Attached FleetAutoScaler %s to fleet %s", fas.ObjectMeta.Name, fas.Spec.FleetName)
```

Comment:
Is there a reason we are doing this twice?

Also, I don't think "CreatingFleetAutoScaler" needs to be cast to string, it already is a string?

Reply:
I don't want to do it twice, hence the Q.
What I wanted is to push the event in both the fleet and fleetautoscaler queues.

Agreed with string conversion, don't know who put it there :D

Reply:
You know, I totally missed that you are pushing to both :man_facepalming: yeah. There is no special way to do that. You did the right thing.

Reply:
Added a comment explaining the reason for duplication